cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,288 @@
+func (fc *FirewallController) Run(stopCh <-chan struct{}) {
+	ctx, cancel := context.WithTimeout(context.Background(), serviceSyncPeriod)
+	defer cancel()
+
+	wait.Until(func() {
+		targetFirewallState := fc.fwCache.firewall.state
+		fm := *fc.firewallManager
+
+		currentFirewallState, err := fm.Get(context.Background())
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if &currentFirewallState != nil {
+			if cmp.Equal(currentFirewallState, targetFirewallState) {
+				return
+			}
+		}
+		err = fc.onServiceChange(ctx)
+		if err != nil {
+			klog.Errorf("Failed to reconcile worker firewall: %s", err)
+		}
+
+	}, 5*time.Minute, stopCh)
+}
+
+// Get returns the current CCM worker firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	if notNil(fm) {
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fm.fwCache.firewall.state.ID)
+		if err != nil {
+			klog.Errorf("failed to retrieve firewall by ID from local cache")
+		}
+		if resp != nil && resp.StatusCode == 404 {
+			klog.Infof("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			return fw, nil
+		}
+	}
+
+	// iterate through firewall API provided list and return the firewall with the matching firewall name.
+	firewallList, err := allFirewallList(ctx, fm.client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve firewall from DO firewall API: %s", err)
+	}
+	formattedWorkerFirewallName := fmt.Sprintf(firewallWorkerCCMNameFormat, fm.workerFirewallName)
+	for _, fw := range firewallList {
+		if fw.Name == formattedWorkerFirewallName {
+			return &fw, nil
+		}
+	}
+	return nil, nil
+}
+
+func notNil(fm *firewallManagerOp) bool {
+	if fm.fwCache == nil || fm.fwCache.firewall == nil || fm.fwCache.firewall.state == nil {
+		return false
+	}
+	return true
+}
+
+// Set applies the given inbound rules to the CCM worker firewall when the current rules and target rules differ.
+func (fm *firewallManagerOp) Set(ctx context.Context, svcInboundRules []godo.InboundRule) error {
+	fm.fwCache.mu.RLock()
+	defer fm.fwCache.mu.RUnlock()
+
+	targetFirewall := fm.fwCache.firewall.state
+	currentFirewall, err := fm.Get(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get firewall state: %s", err)
+	}
+	// if the current firewall state retrieved from the DO firewall API does not match,
+	// we'll need to reconcile the delta between the current firewall (DO API) and the
+	// target firewall (locally cached CCM firewall).
+	if !cmp.Equal(targetFirewall.InboundRules, svcInboundRules) {
+		// update api firewall inbound rules with the expected inbound rules.
+		fwSvc := fm.client.Firewalls
+		fr := &godo.FirewallRequest{
+			Name:          currentFirewall.Name,
+			InboundRules:  svcInboundRules,
+			OutboundRules: currentFirewall.OutboundRules,
```

Comment:
empty or omitted @timoreimann ?

Reply:
I think either should be fine since [the `OutboundRules` field on the godo struct does not carry an `omitempty` tag](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/cb072debb27737442976bdb704b7e7b0e8f4412d/vendor/github.com/digitalocean/godo/firewalls.go#L60). Would be great though if you could double-check by changing a CCM-managed firewall to have some (non-empty) outbound rules and restarting CCM to let it reconcile the change away. If everything goes as desired, the outbound rules should be cleared afterwards.